### PR TITLE
Add prototype tests for minimum resource limits

### PR DESCRIPTION
**Pull request by JusSn**
WIP. I'm sure there're other things I need to do in order to finish this out.

### CODE REVIEW

File: src/suites/cts/validation/createBindGroupLayout.spec.ts [hunk 1]
```diff
@@ -148,3 +148,91 @@ g.test('dynamic set to true is allowed only for buffers', async t => {
...
 ]);
+
+g.test('number of resources per stage exceeds maximum value for resource type', async t => {
+  const { type, maximumCount } = t.params;
+
+  const maxResourceBindings: GPUBindGroupLayoutBinding[] = [];
+  for (let i = 0; i < maximumCount; i++) {
+    maxResourceBindings.push({
+      binding: i,
+      visibility: GPUShaderStage.FRAGMENT,
```
Comment (kainino0x) [inline src/suites/cts/validation/createBindGroupLayout.spec.ts]: To make this a little more complete by testing different stages, how about this: Here, use `VERTEX|FRAGMENT|COMPUTE`. Then...

File: src/suites/cts/validation/createBindGroupLayout.spec.ts [hunk 2]
```diff
@@ -148,3 +148,91 @@ g.test('dynamic set to true is allowed only for buffers', async t => {
...
+
+  const goodDescriptor: GPUBindGroupLayoutDescriptor = { bindings: maxResourceBindings };
+
+  // Control
+  t.device.createBindGroupLayout(goodDescriptor);
+
+  const badDescriptor = clone(goodDescriptor);
+  badDescriptor.bindings.push({
+    binding: maximumCount,
+    visibility: GPUShaderStage.FRAGMENT,
```
Comment (kainino0x) [inline src/suites/cts/validation/createBindGroupLayout.spec.ts]: ... here, use VERTEX, FRAGMENT, or COMPUTE based on a parameter.

You can easily add this to the existing params list using `pcombine` (cartesian product) with `poptions('visibility', ['VERTEX', 'FRAGMENT', 'COMPUTE']`.

File: src/suites/cts/validation/createBindGroupLayout.spec.ts [hunk 3]
```diff
@@ -148,3 +148,91 @@ g.test('dynamic set to true is allowed only for buffers', async t => {
...
+    t.device.createBindGroupLayout(badDescriptor);
+  });
+}).params([
+  { type: 'uniform-buffer', maximumCount: 12 },
+  { type: 'storage-buffer', maximumCount: 4 },
+  { type: 'readonly-storage-buffer', maximumCount: 4 },
+  { type: 'sampler', maximumCount: 16 },
+  { type: 'sampled-texture', maximumCount: 16 },
+  { type: 'storage-texture', maximumCount: 4 },
+]);
```
Comment (kainino0x) [inline src/suites/cts/validation/createBindGroupLayout.spec.ts]: Another idea for a test, maybe later: create maximumCount of one type, and then add one more of a different type. In all cases except storage-buffer + readonly-storage-buffer, this should pass validation.

**Comment by kainino0x** [comment]
btw, there are a few typescript errors in the build log. Try `npx grunt pre` to check this locally.

**Comment by kainino0x** [comment]
I'm working on updating this PR. It turns out there's an issue with the way the typescript types for the bitflags, which makes it difficult.

**Comment by kainino0x** [comment]
@JusSn: please review the 2 commits I pushed on top of yours

**Comment by kainino0x** [comment]
When merging with master, I did a bit more refactoring and ended up:
- realizing the validation should happen on pipeline layout creation, not bind group creation (I think - correct me if I'm wrong), so fixed tests for that
- overengineering the test case list...

PTAL @JusSn + @austinEng 
(cc @Kangz)

File: src/suites/cts/validation/createBindGroupLayout.spec.ts [hunk 4]
```diff
@@ -111,17 +107,121 @@ g.test('dynamic set to true is allowed only for buffers', async t => {
...
+    bindings: [{ binding: 0, visibility: GPUShaderStage.FRAGMENT, type, hasDynamicOffset: true }],
   };
 
   t.expectValidationError(() => {
     t.device.createBindGroupLayout(descriptor);
   }, !success);
 }).params(poptions('type', bindingTypes));
+
+{
+  function pickExtraBindingTypes(type: GPUBindingType, extraTypeSame: boolean): GPUBindingType[] {
```
Comment (austinEng) [inline src/suites/cts/validation/createBindGroupLayout.spec.ts]: nit: Please add comments about what this function is doing and why it helps us in the test.

From what I understand, we want to test cases when an additional binding pushes us over a limit, and when an additional binding does not push us over (`extraTypeSame`). If it's a storage buffer, then either types of storage buffer count toward the limit, ..., etc.

File: src/suites/cts/validation/createBindGroupLayout.spec.ts [hunk 5]
```diff
@@ -111,17 +107,121 @@ g.test('dynamic set to true is allowed only for buffers', async t => {
...
+          return ['storage-buffer', 'readonly-storage-buffer'];
+        default:
+          return [type];
+      }
+    } else {
+      return type === 'sampler' ? ['sampled-texture'] : ['sampler'];
+    }
+  }
+
+  const kCasesForMaxResourcesPerStageTests: ParamSpec[] = [];
```
Comment (austinEng) [inline src/suites/cts/validation/createBindGroupLayout.spec.ts]: nit: This could be declared outside of the { } scope, as well as the tests. This helps the tests be indented the same, and hides `pickExtraBindingTypes` when it's strictly used to populate `kCasesForMaxResourcesPerStageTests`

File: src/suites/cts/validation/createBindGroupLayout.spec.ts [hunk 6]
```diff
@@ -111,17 +107,121 @@ g.test('dynamic set to true is allowed only for buffers', async t => {
...
+    const goodLayout = t.device.createBindGroupLayout({ bindings: maxResourceBindings });
+
+    // Control
+    t.device.createPipelineLayout({ bindGroupLayouts: [goodLayout] });
+
+    const extraLayout = t.device.createBindGroupLayout({
+      bindings: [{ binding: 0, visibility: extraVisibility, type: extraType }],
+    });
+
+    const hasCollision =
```
Comment (austinEng) [inline src/suites/cts/validation/createBindGroupLayout.spec.ts]: nit: Comment to explain what a "collision" is and why it's an error, and why they must be storage buffers to be regarded as colliding.

File: src/suites/cts/validation/createBindGroupLayout.spec.ts [hunk 7]
```diff
@@ -3,10 +3,14 @@ createBindGroupLayout validation tests.
 `;
 
 import { C, TestGroup, poptions } from '../../../framework/index.js';
-import { bindingTypeInfo, bindingTypes } from '../format_info.js';
+import { ParamSpec } from '../../../framework/params/index.js';
+import { bindingTypeInfo, bindingTypes, shaderStages } from '../format_info.js';
```
Comment (austinEng) [inline src/suites/cts/validation/createBindGroupLayout.spec.ts]: Maybe not for this PR, but it would be nice if these constant exports from format_info had different casing. Multiple times in this review I was looking for local definitions of `bindingTypes` or `shaderStages` . Maybe `kBindingTypeInfo`, `kBindingTypes`, and `kShaderStages` would be more clear.

**Comment by kainino0x** [comment]
PTAAL (probably want to `ignore whitespace changes`)

File: src/suites/cts/validation/createBindGroupLayout.spec.ts [hunk 8]
```diff
@@ -3,10 +3,19 @@ createBindGroupLayout validation tests.
...
+  kBindingTypeInfo,
+  kBindingTypes,
+  kPerStageBindingLimits,
+  kShaderStages,
+} from '../capability_info.js';
 
 import { ValidationTest } from './validation_test.js';
 
+// TODO: Move this somewhere central?
+const kMaxBindingsPerBindGroup = 16;
```
Comment (austinEng) [inline src/suites/cts/validation/createBindGroupLayout.spec.ts]: `capability_info` seems like a pretty reasonable place now

File: src/suites/cts/capability_info.ts [hunk 9]
```diff
@@ -54,25 +54,41 @@ export const textureFormatInfo: {
...
   };
 } = /* prettier-ignore */ {
-  'uniform-buffer':          { type: 'buffer',  isStorageBuffer: false, maxPerShaderStage: 12, maxDynamicCount: 8 },
-  'storage-buffer':          { type: 'buffer',  isStorageBuffer: true,  maxPerShaderStage:  4, maxDynamicCount: 4 },
-  'readonly-storage-buffer': { type: 'buffer',  isStorageBuffer: true,  maxPerShaderStage:  4, maxDynamicCount: 4 },
-  'sampler':                 { type: 'sampler', isStorageBuffer: false, maxPerShaderStage: 16, maxDynamicCount: 0 },
-  'sampled-texture':         { type: 'texture', isStorageBuffer: false, maxPerShaderStage: 16, maxDynamicCount: 0 },
-  'storage-texture':         { type: 'texture', isStorageBuffer: false, maxPerShaderStage:  4, maxDynamicCount: 0 },
+  'uniform-buffer':          { type: 'buffer',  validStages: 7, perStageLimitType: 'uniform-buffer',  maxDynamicCount: 8 },
+  'storage-buffer':          { type: 'buffer',  validStages: 6, perStageLimitType: 'storage-buffer',  maxDynamicCount: 4 },
```
Comment (austinEng) [inline src/suites/cts/capability_info.ts]: Not a huge fan of `6` because it's rather opaque. Consider at least `2 | 4`, or even define `kShaderStageVertex = ...` etc. and do `kShaderStageFragment | kShaderStageCompute`
